pallets/phala/src/ott.rs

```diff
@@ -0,0 +1,186 @@
+//! transferring again in the lifetime.
+//!
+//! This pallet is useful when token transfer is disabled due to security concerns in the
+//! typical "progressive launch process" of Substrate-based blockchains, while users still need
+//! subaccounts to pay the transaction fee for the already enabled features (e.g. Phala mining).
+
+pub use self::pallet::*;
+
+#[frame_support::pallet]
+pub mod pallet {
+	use frame_support::{
+		dispatch::DispatchResult,
+		pallet_prelude::*,
+		traits::{Currency, ExistenceRequirement::KeepAlive, StorageVersion},
+		transactional,
+	};
+	use frame_system::pallet_prelude::*;
+	use sp_std::vec::Vec;
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+
+		type Currency: Currency<Self::AccountId>;
+	}
+
+	const STORAGE_VERSION: StorageVersion = StorageVersion::new(1);
+
+	#[pallet::pallet]
+	#[pallet::generate_store(pub(super) trait Store)]
+	#[pallet::storage_version(STORAGE_VERSION)]
+	pub struct Pallet<T>(_);
+
+	/// The accounts forbidden to transfer
+	#[pallet::storage]
+	pub type BlacklistedAccounts<T: Config> = StorageMap<_, Twox64Concat, T::AccountId, ()>;
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	#[pallet::metadata(Vec<T::AccountId> = "Vec<AccountId>")]
+	pub enum Event<T: Config> {
+		AccountsBlacklisted(Vec<T::AccountId>),
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// Operation not permitted because the sender is already in the blacklist
+		SenderAlreadyBlacklisted,
+		/// Operation not permitted because one of the destination is already in the blacklist
+		DestinationAlreadyBlacklisted,
+	}
+
+	type BalanceOf<T> =
+		<<T as Config>::Currency as Currency<<T as frame_system::Config>::AccountId>>::Balance;
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Distributes some amounts to each specified accounts and mark the sender and destination
+		/// accounts as blacklisted.
+		#[pallet::weight(0)]
+		#[transactional]
+		pub fn distribute(
+			origin: OriginFor<T>,
+			transfers: Vec<(T::AccountId, BalanceOf<T>)>,
+		) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+			// Check blacklist
+			ensure!(
+				BlacklistedAccounts::<T>::get(&who) == None,
+				Error::<T>::SenderAlreadyBlacklisted
+			);
+			for (dest, _) in &transfers {
+				ensure!(
+					BlacklistedAccounts::<T>::get(dest) == None,
+					Error::<T>::DestinationAlreadyBlacklisted
+				);
+			}
+			// Try to transfer (the entire call will be rolledback if there's no enough funds)
+			let mut blacklisted = Vec::<T::AccountId>::new();
+			for (dest, amount) in &transfers {
+				T::Currency::transfer(&who, dest, *amount, KeepAlive)?;
+				BlacklistedAccounts::<T>::insert(dest, ());
+				blacklisted.push(dest.clone());
+			}
+			BlacklistedAccounts::<T>::insert(&who, ());
+			blacklisted.push(who);
+
+			Self::deposit_event(Event::<T>::AccountsBlacklisted(blacklisted));
```

Comment:
I think too much writing operations here, how about

```
let blacklisted: Vec<T::AccountId> = transfers.iter().map(|t| t.0).collect();
blacklisted.push(who);
Self::deposit_event(Event::<T>::AccountsBlacklisted(blacklisted);
```

Reply:
It's an in-memory vec. Should be fine.